### PR TITLE
55 h5stats

### PR DESCRIPTION
**Pull request by emlynjdavies**
addresses #50

### CODE REVIEW

File: pysilcam/tests/test_process.py [hunk 1]
```diff
@@ -96,28 +96,38 @@ def test_output_files():
...
-    column_string += ',export name,timestamp,saturation\n'
+        column_string.append('probability_' + c)
+
+    matching_elements = list(
+            set(column_string) &
+            set(stats.columns.tolist()))
 
     # check that output STATS file contains expected columns
-    assert lines[0] == column_string
+    assert len(matching_elements) == len(stats.columns.tolist()) == len(column_string), 'output STATS file contains unexpected columns'
```
Comment (tankefugl) [inline pysilcam/tests/test_process.py:128]: E501: line too long (135 > 130 characters)

File: pysilcam/silcamgui/silcamgui.py [hunk 2]
```diff
@@ -822,9 +822,9 @@ def stop_record(self):
 
         def load_stats_filename(self):
             self.stats_filename = QFileDialog.getOpenFileName(self,
-                    caption = 'Load a *-STATS.csv file',
+                    caption = 'Load a *-STATS.h5 file',
                     directory = self.datadir,
-                    filter = (('*-STATS.csv'))
+                    filter = (('*-STATS.h5'))
```
Comment (tankefugl) [inline pysilcam/silcamgui/silcamgui.py:827]: E251: unexpected spaces around keyword / parameter equals

Comment (tankefugl) [inline pysilcam/silcamgui/silcamgui.py:827]: E251: unexpected spaces around keyword / parameter equals

Comment (tankefugl) [inline pysilcam/silcamgui/silcamgui.py:827]: E128: continuation line under-indented for visual indent

File: pysilcam/silcamgui/silcamgui.py [hunk 3]
```diff
@@ -822,9 +822,9 @@ def stop_record(self):
 
         def load_stats_filename(self):
             self.stats_filename = QFileDialog.getOpenFileName(self,
-                    caption = 'Load a *-STATS.csv file',
+                    caption = 'Load a *-STATS.h5 file',
```
Comment (tankefugl) [inline pysilcam/silcamgui/silcamgui.py:825]: E251: unexpected spaces around keyword / parameter equals

Comment (tankefugl) [inline pysilcam/silcamgui/silcamgui.py:825]: E251: unexpected spaces around keyword / parameter equals

Comment (tankefugl) [inline pysilcam/silcamgui/silcamgui.py:825]: E128: continuation line under-indented for visual indent

File: pysilcam/silcamgui/silcamgui.py [hunk 4]
```diff
@@ -555,7 +555,7 @@ def export_summary_data_slow(self):
             labs = [l.get_label() for l in lns]
             plt.legend(lns, labs)
 
-            plt.savefig(self.stats_filename.replace('-STATS.csv','') +
+            plt.savefig(self.stats_filename.replace('-STATS.h5','') +
```
Comment (tankefugl) [inline pysilcam/silcamgui/silcamgui.py:558]: E231: missing whitespace after ','

File: pysilcam/silcamgui/silcamgui.py [hunk 5]
```diff
@@ -457,23 +457,23 @@ def export_summary_figure(self):
...
+            plt.savefig(self.stats_filename.replace('-STATS.h5','') + '-Summary_oil.png',
                     dpi=600, bbox_inches='tight')
 
             plt.figure(figsize=(20,12))
             self.status_update('Creating summary figure (gas)....')
             scplt.summarise_fancy_stats(self.stats_filename,
                     self.configfile, monitor=False, oilgas=scpp.outputPartType.gas)
             self.status_update('Saving summary figure (gas)....')
-            plt.savefig(self.stats_filename.replace('-STATS.csv','') + '-Summary_gas.png',
+            plt.savefig(self.stats_filename.replace('-STATS.h5','') + '-Summary_gas.png',
```
Comment (tankefugl) [inline pysilcam/silcamgui/silcamgui.py:476]: E231: missing whitespace after ','

File: pysilcam/silcamgui/silcamgui.py [hunk 6]
```diff
@@ -457,23 +457,23 @@ def export_summary_figure(self):
...
+            plt.savefig(self.stats_filename.replace('-STATS.h5','') + '-Summary.png',
                     dpi=600, bbox_inches='tight')
 
             plt.figure(figsize=(20,12))
             self.status_update('Creating summary figure (oil)....')
             scplt.summarise_fancy_stats(self.stats_filename,
                     self.configfile, monitor=False, oilgas=scpp.outputPartType.oil)
             self.status_update('Saving summary figure (oil)....')
-            plt.savefig(self.stats_filename.replace('-STATS.csv','') + '-Summary_oil.png',
+            plt.savefig(self.stats_filename.replace('-STATS.h5','') + '-Summary_oil.png',
```
Comment (tankefugl) [inline pysilcam/silcamgui/silcamgui.py:468]: E231: missing whitespace after ','

File: pysilcam/silcamgui/silcamgui.py [hunk 7]
```diff
@@ -457,23 +457,23 @@ def export_summary_figure(self):
             scplt.summarise_fancy_stats(self.stats_filename,
                     self.configfile, monitor=False)
             self.status_update('Saving summary figure (all)....')
-            plt.savefig(self.stats_filename.replace('-STATS.csv','') + '-Summary.png',
+            plt.savefig(self.stats_filename.replace('-STATS.h5','') + '-Summary.png',
```
Comment (tankefugl) [inline pysilcam/silcamgui/silcamgui.py:460]: E231: missing whitespace after ','

File: pysilcam/silcamgui/interactive_summary.py [hunk 8]
```diff
@@ -559,7 +560,7 @@ def update_plot(self, save=False):
 
         if save:
             timestring = pd.to_datetime(psd_start[0]).strftime('D%Y%m%dT%H%M%S')
-            outputname = self.stats_filename.replace('-STATS.csv','-PSD-' + timestring)
+            outputname = self.stats_filename.replace('-STATS.h5','-PSD-' + timestring)
```
Comment (tankefugl) [inline pysilcam/silcamgui/interactive_summary.py:563]: E231: missing whitespace after ','

File: pysilcam/silcamgui/interactive_summary.py [hunk 9]
```diff
@@ -342,7 +342,8 @@ def load_from_timeseries(self):
 
     def load_from_stats(self):
         '''loads stats data and converts to timeseries without saving'''
-        stats = pd.read_csv(self.stats_filename, parse_dates=['timestamp'])
+        stats= pd.read_hdf(self.stats_filename, 'ParticleStats/stats')
```
Comment (tankefugl) [inline pysilcam/silcamgui/interactive_summary.py:345]: E225: missing whitespace around operator

File: pysilcam/silcamgui/guicalcs.py [hunk 10]
```diff
@@ -118,7 +118,7 @@ def export_timeseries(configfile, statsfile):
         gor.append(np.sum(vdts_av_gas)/np.sum(vdts_av_oil))
 
     outpath, outfile = os.path.split(statsfile)
-    outfile = outfile.replace('-STATS.csv','')
+    outfile = outfile.replace('-STATS.h5','')
```
Comment (tankefugl) [inline pysilcam/silcamgui/guicalcs.py:121]: E231: missing whitespace after ','

File: pysilcam/postprocess.py [hunk 11]
```diff
@@ -935,10 +935,10 @@ def trim_stats(stats_csv_file, start_time, end_time, write_new=False, stats=[]):
     actual_start = pd.to_datetime(trimmed_stats['timestamp'].min()).strftime('D%Y%m%dT%H%M%S.%f')
     actual_end = pd.to_datetime(trimmed_stats['timestamp'].max()).strftime('D%Y%m%dT%H%M%S.%f')
 
-    path, name = os.path.split(stats_csv_file)
+    path, name = os.path.split(stats_file)
 
-    outname = os.path.join(path, name.replace('-STATS.csv','')) + '-Start' + str(actual_start) + '-End' + str(
-        actual_end) + '-STATS.csv'
+    outname = os.path.join(path, name.replace('-STATS.h5','')) + '-Start' + str(actual_start) + '-End' + str(
```
Comment (tankefugl) [inline pysilcam/postprocess.py:940]: E231: missing whitespace after ','

File: pysilcam/postprocess.py [hunk 12]
```diff
@@ -897,29 +897,29 @@ def stats_to_xls_png(config_file, stats_filename, oilgas=outputPartType.all):
     timestamp = np.min(pd.to_datetime(df['Time']))
     dfa['Time'] = timestamp
     
-    dfa.to_excel(stats_filename.replace('-STATS.csv','') +
+    dfa.to_excel(stats_filename.replace('-STATS.h5','') +
```
Comment (tankefugl) [inline pysilcam/postprocess.py:900]: E231: missing whitespace after ','

File: pysilcam/postprocess.py [hunk 13]
```diff
@@ -878,7 +878,7 @@ def stats_to_xls_png(config_file, stats_filename, oilgas=outputPartType.all):
 
     df = make_timeseries_vd(stats, settings)
 
-    df.to_excel(stats_filename.replace('-STATS.csv','') +
+    df.to_excel(stats_filename.replace('-STATS.h5','') +
```
Comment (tankefugl) [inline pysilcam/postprocess.py:881]: E231: missing whitespace after ','

**Review by tankefugl — CHANGES_REQUESTED** [review]
[Lintly](https://github.com/grantmcconnaughey/Lintly) has detected code quality issues in this pull request.

File: pysilcam/silcamgui/silcamgui.py (hunk 4 above)
Comment (tankefugl) [inline pysilcam/silcamgui/silcamgui.py:558]: E231: missing whitespace after ','

File: pysilcam/silcamgui/silcamgui.py (hunk 5 above)
Comment (tankefugl) [inline pysilcam/silcamgui/silcamgui.py:476]: E231: missing whitespace after ','

File: pysilcam/silcamgui/silcamgui.py (hunk 6 above)
Comment (tankefugl) [inline pysilcam/silcamgui/silcamgui.py:468]: E231: missing whitespace after ','

File: pysilcam/silcamgui/silcamgui.py (hunk 7 above)
Comment (tankefugl) [inline pysilcam/silcamgui/silcamgui.py:460]: E231: missing whitespace after ','

File: pysilcam/silcamgui/interactive_summary.py (hunk 8 above)
Comment (tankefugl) [inline pysilcam/silcamgui/interactive_summary.py:563]: E231: missing whitespace after ','

**Review by tankefugl — CHANGES_REQUESTED** [review]
[Lintly](https://github.com/grantmcconnaughey/Lintly) has detected code quality issues in this pull request.

File: pysilcam/silcamgui/silcamgui.py [hunk 14]
```diff
@@ -257,12 +255,12 @@ def fillInConfigEditor(self, inputFile):
...
 
         for port in listPortCom:
-            if (port != self.settings.PostProcess.com_port):
+            if port != self.settings.PostProcess.com_port:
                 self.ui.com_portEdit.addItem(port)
 
         self.ui.window_sizeEdit.setText(str(self.settings.PostProcess.window_size))
 
-        if (self.settings.ExportParticles.export_images == True):
+        if self.settings.ExportParticles.export_images == True:
```
Comment (tankefugl) [inline pysilcam/silcamgui/silcamgui.py:263]: E712: comparison to True should be 'if cond is True:' or 'if cond:'

**Review by tankefugl — DISMISSED** [review]
[Lintly](https://github.com/grantmcconnaughey/Lintly) has detected code quality issues in this pull request.

**Comment by emlynjdavies** [comment]
hangs on windows

**Comment by emlynjdavies** [comment]
On windows:

Traceback (most recent call last):
  File "C:\Users\emlynd\Anaconda3\envs\silcam\Scripts\silcam-script.py", line 11, in <module>
    load_entry_point('PySilCam', 'console_scripts', 'silcam')()
  File "c:\users\emlynd\pysilcam\pysilcam\__main__.py", line 97, in silcam
    nbImages=nbImages, overwriteSTATS=overwriteSTATS)
  File "c:\users\emlynd\pysilcam\pysilcam\__main__.py", line 319, in silcam_process
    settings, rts=rts)
  File "c:\users\emlynd\pysilcam\pysilcam\__main__.py", line 581, in collector
    writeCSV(datafilename, task)
  File "c:\users\emlynd\pysilcam\pysilcam\__main__.py", line 623, in writeCSV
    with pd.HDFStore(datafilename + '-STATS.h5', 'a') as fh:
  File "C:\Users\emlynd\Anaconda3\envs\silcam\lib\site-packages\pandas\io\pytables.py", line 450, in __init__
    'importing'.format(ex=str(ex)))
ImportError: HDFStore requires PyTables, "DLL load failed: The specified module could not be found." problem importing

**Review by nepstad — DISMISSED** [review]

File: pysilcam/silcamgui/silcamgui.py [hunk 15]
```diff
@@ -31,6 +29,7 @@
 IP = '192.168.1.2'
 DEFAULT_CONFIG = os.path.join(os.path.dirname(__file__), '../config_example.ini')
```
Comment (nepstad) [inline pysilcam/silcamgui/silcamgui.py:30]: Avoid using / in path construction

File: pysilcam/tests/test_standards.py [hunk 16]
```diff
@@ -37,7 +37,7 @@ def test_big_standards():
     with open(conf_file_out, 'w') as conf_file_hand:
         conf.write(conf_file_hand)
 
-    stats_file = os.path.join(ROOTPATH, 'STANDARDS', 'proc', 'StandardsBig-STATS.csv')
+    stats_file = os.path.join(ROOTPATH, 'STANDARDS/proc/StandardsBig-STATS.h5')
```
Comment (nepstad) [inline pysilcam/tests/test_standards.py:40]: Avoid using / in path construction

File: pysilcam/tests/test_standards.py [hunk 17]
```diff
@@ -79,7 +73,7 @@ def test_small_standards():
     with open(conf_file_out, 'w') as conf_file_hand:
         conf.write(conf_file_hand)
 
-    stats_file = os.path.join(ROOTPATH, 'STANDARDS', 'proc', 'StandardsSmall-STATS.csv')
+    stats_file = os.path.join(ROOTPATH, 'STANDARDS/proc/StandardsSmall-STATS.h5')
```
Comment (nepstad) [inline pysilcam/tests/test_standards.py:76]: Avoid using / in path construction

File: tfl_tools/investigate_particles.py [hunk 18]
```diff
@@ -31,15 +31,15 @@ def particle_generator():
     maxdepth = 70
 
     config_file = '/mnt/ARRAY/ENTICE/Data/configs/config.ini'
-    stats_csv_file = '/mnt/ARRAY/ENTICE/Data/proc/' + stn + '-STATS.csv'
+    stats_file = '/mnt/ARRAY/ENTICE/Data/proc/' + stn + '-STATS.h5'
```
Comment (nepstad) [inline tfl_tools/investigate_particles.py:34]: Avoid using / in path construction

File: tfl_tools/silcam_classifier_database_builder.py [hunk 19]
```diff
@@ -19,7 +19,7 @@
 
 DATABASE_PATH = '/mnt/ARRAY/silcam_classification_database'
 config_file = '/mnt/ARRAY/ENTICE/Data/configs/config.ini'
-stats_csv_file = '/mnt/ARRAY/ENTICE/Data/proc/STN10-STATS.csv'
+stats_file = '/mnt/ARRAY/ENTICE/Data/proc/STN10-STATS.h5'
```
Comment (nepstad) [inline tfl_tools/silcam_classifier_database_builder.py:22]: Avoid using / in path construction

File: tfl_tools/silcam_self_learn.py [hunk 20]
```diff
@@ -9,7 +9,7 @@
 
 DATABASE_PATH = '/mnt/ARRAY/silcam_classification_database'
 config_file = '/mnt/nasdrive/Miljoteknologi/MK102013220_SILCAM_IPR_EJD/RUNDE_June2017/LowMag/config_LowRes.ini'
-stats_csv_file = '/mnt/nasdrive/Miljoteknologi/MK102013220_SILCAM_IPR_EJD/RUNDE_June2017/LowMag/proc/STN05-STATS.csv'
+stats_file = '/mnt/nasdrive/Miljoteknologi/MK102013220_SILCAM_IPR_EJD/RUNDE_June2017/LowMag/proc/STN05-STATS.h5'
```
Comment (nepstad) [inline tfl_tools/silcam_self_learn.py:12]: Avoid using / in path construction

**Comment by emlynjdavies** [comment]
pysilcam\tests\test_standards.py:89:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

path_or_buf = 'C:\\Users\\emlynd\\Documents\\pysilcam-testdata\\pysilcam-testdata\\unittest-data\\STANDARDS\\proc\\StandardsSmall-STATS.h5'
key = 'ParticleStats', mode = 'stats', errors = 'strict', where = None, start = None
stop = None, columns = None, iterator = False, chunksize = None, kwargs = {}

    def read_hdf(
        path_or_buf,
        key=None,
        mode: str = "r",
        errors: str = "strict",
        where=None,
        start: Optional[int] = None,
        stop: Optional[int] = None,
        columns=None,
        iterator=False,
        chunksize: Optional[int] = None,
        **kwargs,
    ):
        """
        Read from the store, close it if we opened it.

        Retrieve pandas object stored in file, optionally based on where
        criteria.

        .. warning::

           Pandas uses PyTables for reading and writing HDF5 files, which allows
           serializing object-dtype data with pickle when using the "fixed" format.
           Loading pickled data received from untrusted sources can be unsafe.

           See: https://docs.python.org/3/library/pickle.html for more.

        Parameters
        ----------
        path_or_buf : str, path object, pandas.HDFStore or file-like object
            Any valid string path is acceptable. The string could be a URL. Valid
            URL schemes include http, ftp, s3, and file. For file URLs, a host is
            expected. A local file could be: ``file://localhost/path/to/table.h5``.

            If you want to pass in a path object, pandas accepts any
            ``os.PathLike``.

            Alternatively, pandas accepts an open :class:`pandas.HDFStore` object.

            By file-like object, we refer to objects with a ``read()`` method,
            such as a file handler (e.g. via builtin ``open`` function)
            or ``StringIO``.
        key : object, optional
            The group identifier in the store. Can be omitted if the HDF file
            contains a single pandas object.
        mode : {'r', 'r+', 'a'}, default 'r'
            Mode to use when opening the file. Ignored if path_or_buf is a
            :class:`pandas.HDFStore`. Default is 'r'.
        errors : str, default 'strict'
            Specifies how encoding and decoding errors are to be handled.
            See the errors argument for :func:`open` for a full list
            of options.
        where : list, optional
            A list of Term (or convertible) objects.
        start : int, optional
            Row number to start selection.
        stop  : int, optional
            Row number to stop selection.
        columns : list, optional
            A list of columns names to return.
        iterator : bool, optional
            Return an iterator object.
        chunksize : int, optional
            Number of rows to include in an iteration when using an iterator.
        **kwargs
            Additional keyword arguments passed to HDFStore.

        Returns
        -------
        item : object
            The selected object. Return type depends on the object stored.

        See Also
        --------
        DataFrame.to_hdf : Write a HDF file from a DataFrame.
        HDFStore : Low-level access to HDF files.

        Examples
        --------
        >>> df = pd.DataFrame([[1, 1.0, 'a']], columns=['x', 'y', 'z'])
        >>> df.to_hdf('./store.h5', 'data')
        >>> reread = pd.read_hdf('./store.h5')
        """
        if mode not in ["r", "r+", "a"]:
>           raise ValueError(
                f"mode {mode} is not allowed while performing a read. "
                f"Allowed modes are r, r+ and a."
            )
E           ValueError: mode stats is not allowed while performing a read. Allowed modes are r, r+ and a.

..\Anaconda3\envs\silcam\lib\site-packages\pandas\io\pytables.py:359: ValueError
------------------------------ Captured stdout call -------------------------------

File: tfl_tools/silcam_self_learn.py (hunk 20 above)
Comment (emlynjdavies) [reply tfl_tools/silcam_self_learn.py:12]: #79

File: tfl_tools/silcam_classifier_database_builder.py (hunk 19 above)
Comment (emlynjdavies) [reply tfl_tools/silcam_classifier_database_builder.py:22]: refer to #79

File: tfl_tools/investigate_particles.py (hunk 18 above)
Comment (emlynjdavies) [reply tfl_tools/investigate_particles.py:34]: refer to #78

**Review by nepstad — APPROVED** [review]

**Review by WillRNaylor — APPROVED** [review]
I just read the diff, and didn't test locally. Will test locally if you want me to, or you can just merge yourself.

Nice to see all the formatting tidying too.

**Pull request merged**